File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import (
	"errors"
)

func CollatzConjecture(n int) (int, error) {

	if n <= 0 {
		return 0, errors.New("n is 0 or negative")
	}

	steps := 0

	if n == 1 {
		return steps, nil
	}

	for n != 1 {

		if n%2 == 0 {
			n /= 2
		} else {
			n = 3*n + 1
		}

		steps++

	}

	return steps, nil
}

File: scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
import scala.annotation.tailrec

object CollatzConjecture {
  def steps(start: Int): Option[Int] = {
    if (start <= 0) None
    else {
      Some(calculateSteps(0, start))
    }
  }

  @tailrec
  def calculateSteps(steps: Int, start: Int): Int = {

    if (start == 1) steps
    else if (start % 2 == 0)
      calculateSteps((steps + 1), (start / 2))

    else
      calculateSteps((steps + 1), (3 * start + 1))

  }
}